## Step 1

### Dataset Column Details
[Kaggle Link](https://www.kaggle.com/datasets/davinwijaya/employee-turnover) [Original Link](https://edwvb.blogspot.com/2017/10/employee-turnover-how-to-predict-individual-risks-of-quitting.html)
- **stag** - Experience (time)
- **event** - Employee turnover
- **gender** - Employee's gender, female(f), or male(m)
- **age** - Employee's age (year)
- **industry** - Employee's Industry
- **profession** - Employee's profession
- **traffic** - From what pipelene employee came to the company. You contacted the company directly (after learning from advertising, knowing the company's brand, etc.) - advert You contacted the company directly on the recommendation of your friend - NOT an employee of this company-recNErab You contacted the company directly on the recommendation of your friend - an employee of this company - referal You have applied for a vacancy on the job site - youjs The recruiting agency brought you to the employer - KA Invited by the Employer, we knew him before the employment - friends The employer contacted you on the recommendation of a person who knows you - rabrecNErab The employer reached you through your resume on the job site - empjs
- **coach** - Presence of a coach (training) on probation
- **head_gender** - head (supervisor) gender
- **greywage** - The salary does not seem to the tax authorities. Greywage in Russia or Ukraine means that the employer (company) pay
- **way** - Employee's way of transportation
- **extraversion** - Extraversion score
- **independ** - Independend score
- **selfcontrol** - Selfcontrol score
- **anxiety** - Anxiety score
- **novator** - Novator score

In [1]:
import pandas as pd
import plotly.express as px

In [3]:
df = pd.read_csv("turnover.csv", encoding='ISO-8859-1')

In [7]:
df.head()

stag  event gender   age         industry  profession      traffic  \
0   7.030801      1      m  35.0            Banks          HR  rabrecNErab   
1  22.965092      1      m  33.0            Banks          HR        empjs   
2  15.934292      1      f  35.0  PowerGeneration          HR  rabrecNErab   
3  15.934292      1      f  35.0  PowerGeneration          HR  rabrecNErab   
4   8.410678      1      m  32.0           Retail  Commercial        youjs   

  coach head_gender greywage  way  extraversion  independ  selfcontrol  \
0    no           f    white  bus           6.2       4.1          5.7   
1    no           m    white  bus           6.2       4.1          5.7   
2    no           m    white  bus           6.2       6.2          2.6   
3    no           m    white  bus           5.4       7.6          4.9   
4   yes           f    white  bus           3.0       4.1          8.0   

   anxiety  novator  
0      7.1      8.3  
1      7.1      8.3  
2      4.8      8.3  
3      2.5      6.7  
4      7.1      3.7

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1129 entries, 0 to 1128
Data columns (total 16 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   stag          1129 non-null   float64
 1   event         1129 non-null   int64  
 2   gender        1129 non-null   object 
 3   age           1129 non-null   float64
 4   industry      1129 non-null   object 
 5   profession    1129 non-null   object 
 6   traffic       1129 non-null   object 
 7   coach         1129 non-null   object 
 8   head_gender   1129 non-null   object 
 9   greywage      1129 non-null   object 
 10  way           1129 non-null   object 
 11  extraversion  1129 non-null   float64
 12  independ      1129 non-null   float64
 13  selfcontrol   1129 non-null   float64
 14  anxiety       1129 non-null   float64
 15  novator       1129 non-null   float64
dtypes: float64(7), int64(1), object(8)
memory usage: 141.2+ KB


In [10]:
dependent_vars = ["event", "novator"]

In [11]:
protected_class = ["gender", "age", "head_gender"]

- **Dataset**: Employee Turnover
- **Dataset Details**: Predict an Employee's risk of quitting.
- **Regulated Domain**: Employment
- **No. of Observations**: 1129
- **No. of Variables**: 16
- **Dependent Variables**: Event, Novator
- **No of Protected Class Variables**: 3
- **Protected Class Variables**: Gender, Age, and Head Gender.
- **Legal Precedence/law**:
	- **Gender**: Equal Pay Act of 1963; Civil Rights Act of 1964, 1991
	- **Head Gender**: Equal Pay Act of 1963; Civil Rights Act of 1964, 1991
	- **Age**: Age Discrimination in Employment Act of 1967

In [17]:
df["age_is40?"] = df["age"].apply(lambda x: 1 if x<40 else 0)

In [20]:
protected_class = ["gender", "age_is40?", "head_gender"]

In [24]:
for col in protected_class:
    print(col, ":", df[col].unique().tolist())
    print(df[col].value_counts().to_dict())

gender : ['m', 'f']
{'f': 853, 'm': 276}
age_is40? : [1, 0]
{1: 982, 0: 147}
head_gender : ['f', 'm']
{'m': 584, 'f': 545}


In [27]:
def discretize_novator(x):
    if x<5: return "less_than_5"
    if x>=5 and x<8: return "5_to_8"
    if x>=8: return "8+"

In [28]:
df["novator_disc"] = df["novator"].apply(lambda x: discretize_novator(x))

In [29]:
dependent_vars = ["event", "novator_disc"]

In [35]:
for col in dependent_vars:
    print(col, ":", df[col].unique().tolist())
    print(df[col].value_counts().to_dict())

event : [1, 0]
{1: 571, 0: 558}
novator_disc : ['8+', '5_to_8', 'less_than_5']
{'5_to_8': 650, 'less_than_5': 309, '8+': 170}


In [41]:
for p in protected_class:
    for d in dependent_vars:
        fig = px.histogram(df, x=p, color=d)
        fig.update_layout(bargap=0.2)
        fig.show()